test/model/internal_link_test.dart

```diff
@@ -0,0 +1,394 @@
+        ('#narrow/stream/check.2EAPI/',   StreamNarrow(3)),
+        ('#narrow/stream/stream/',        StreamNarrow(4)),
+        ('#narrow/stream/topic/',         StreamNarrow(5)),
+
+        ('#narrow/stream/check.API/',     null),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link, without realm info', () {
+      final stream = eg.stream(name: 'example');
+      final testCases = [
+        ('/#narrow/stream/${stream.name}/', StreamNarrow(stream.streamId)),
+        ('#narrow/stream/${stream.name}/',  StreamNarrow(stream.streamId)),
+      ];
+      checkExpectedNarrows(testCases, streams: [stream]);
+    });
+  });
+
+  group('ParseInternalLink', () {
+    group('topic link parsing', () {
+      final stream = eg.stream(name: "example");
+
+      group('basic', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.streamId}-general/topic/$operand';
+        }
+        final testCases = [
+          (mkUrl('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrl('lunch'),        TopicNarrow(stream.streamId, 'lunch')),
+        ];
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, with dot-encoding', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.name}/topic/$operand';
+        }
+        final testCases = [
+          (mkUrl('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrl('google.2Ecom'), TopicNarrow(stream.streamId, 'google.com')),
+          (mkUrl('google.com'),   null),
+          (mkUrl('topic.20name'), TopicNarrow(stream.streamId, 'topic name')),
+          (mkUrl('stream'),       TopicNarrow(stream.streamId, 'stream')),
+          (mkUrl('topic'),        TopicNarrow(stream.streamId, 'topic')),
+        ];
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, without realm info', () {
+        final testCases = [
+          ('/#narrow/stream/${stream.name}/topic/topic', TopicNarrow(stream.streamId, 'topic')),
+          ('#narrow/stream/${stream.name}/topic/topic',  TopicNarrow(stream.streamId, 'topic')),
+        ];
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+    });
+
+    group('DM link parsing', () {
+      checkExpectedDmNarrow(String testCase) {
+        final expectedNarrow = DmNarrow.withUsers([1, 2],
+          selfUserId: eg.selfUser.userId);
+        checkExpectedNarrows([(testCase, expectedNarrow)], users: [
+          eg.user(userId: 1),
+          eg.user(userId: 2),
+        ]);
+      }
+
+      group('on group PM link', () {
+        checkExpectedDmNarrow('#narrow/dm/1,2-group');
+        checkExpectedDmNarrow('#narrow/pm-with/1,2-group');
+      });
+
+      group('on group PM link including self', () {
+        checkExpectedDmNarrow('#narrow/dm/1,2,${eg.selfUser.userId}-group');
```

Comment:
The zulip-mobile version has a comment that's important context for understanding this test:
```suggestion
      group('on group PM link including self', () {
        // The webapp doesn't generate these, but best to handle them anyway.
        checkExpectedDmNarrow('#narrow/dm/1,2,${eg.selfUser.userId}-group');
```